merge upstream zvol restructuring

ahrens opened the pull request:
Merge commit '5df7e9d85ced424292e8312a7848b6092a29c930' into merge2

merge notes: I moved commit 883275865df4495e31585e4d7465a8eac8142857
"dx4linux: override phys block size" from module/zfs/zvol.c to
module/os/linux/zfs/zvol_os.c

zfs-precommit had some unrelated ZTS failures: http://platform.jenkins.delphix.com/job/devops-gate/job/master/job/zfs-precommit/4832/

running ab-pre-push: http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-orchestrator-pre-push/2259/

This is based on #80.  The unique changes can be seen here: https://github.com/delphix/zfs/commit/26bac0142bbe9cf76ac5cced3499d7b45993b336

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/delphix/zfs?pullRequest=82) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/delphix/zfs?pullRequest=82) before we can accept your contribution.<br/>**1** out of **6** committers have signed the CLA.<br/><br/>:white_check_mark: ahrens<br/>:x: avg-I<br/>:x: mattmacy<br/>:x: freqlabs<br/>:x: KodyKantor<br/>:x: loli10K<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/delphix/zfs?pullRequest=82) it.</sub>

pzakha commented:
> merge notes: I moved commit 8832758
"dx4linux: override phys block size" from module/zfs/zvol.c to
module/os/linux/zfs/zvol_os.c

Should we keep this as a separate commit so that we can keep track of the differences with upstream?

ahrens commented:
@pzakha how would we do that exactly?  I could imagine having a commit that reverts 8832758, then one that merges master, and then one that re-applies 8832758 to the new file.  Not sure that's a great idea since the codebase would be "wrong" at the revert commit.

pzakha commented:
Yeah, I was thinking of something like that, and it does come with the issue you mentioned. Alternatively, we could at least add a comment at
```int zvol_fake_phys_block_size = 1;```
Mentioning that this is a Delphix-only change (this is probably something we should have done since the beginning).

ahrens commented:
@pzakha OK I found a way to do this atomically, by creating a more complicated merge history.  Basically I do the 3 commits (revert, merge, reapply) in one branch and then merge with that (with `--no-ff` to force it to create a merge commit).  The history then looks like:

```
*   commit 24c44899ea36f359fcae4a5ad57b8483d8bfd333 (HEAD -> merge3)
|\  Merge: b0230ebc35 4c3b17f36f
| | Author: Matthew Ahrens <mahrens@delphix.com>
| | Date:   Tue Oct 1 14:26:58 2019 -0700
| |
| |     Merge branch 'merge-test' into merge3
| |
| * commit 4c3b17f36f32ab08f7dea092bbd3134a8713294c (merge-test)
| | Author: Matthew Ahrens <mahrens@delphix.com>
| | Date:   Tue Oct 1 14:25:19 2019 -0700
| |
| |     dx4linux: override phys block size
| |
| *   commit ef62b60df15e6093309113457e2c76f5fb68d6c9
| |\  Merge: 173c8e3bf8 5df7e9d85c
| | | Author: Matthew Ahrens <mahrens@delphix.com>
| | | Date:   Tue Oct 1 14:16:39 2019 -0700
| | |
| | |     Merge commit '5df7e9d85ced424292e8312a7848b6092a29c930' into merge-test
| | |
| | * commit 5df7e9d85ced424292e8312a7848b6092a29c930
| | | Author: Matthew Macy <mmacy@mattmacy.io>
| | | Date:   Wed Sep 25 09:20:30 2019 -0700
| | |
| | |     OpenZFS restructuring - zvol
| | |
| | |     Refactor the zvol in to platform dependent and independent bits.
| | |
| | |     Reviewed-by: Allan Jude <allanjude@freebsd.org>
| | |     Reviewed-by: Jorgen Lundman <lundman@lundman.net>
| | |     Reviewed-by: Igor Kozhukhov <igor@dilos.org>
| | |     Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
| | |     Signed-off-by: Matt Macy <mmacy@FreeBSD.org>
| | |     Closes #9295
| | |
| * | commit 173c8e3bf83712a46f5978d7defe1dce150cb90a
|/ /  Author: Matthew Ahrens <mahrens@delphix.com>
| |   Date:   Tue Oct 1 14:16:18 2019 -0700
| |
| |       Revert "dx4linux: override phys block size"
| |
| |       This reverts commit 883275865df4495e31585e4d7465a8eac8142857.
| |
* |   commit b0230ebc3580ff62827b7a7fbfbb6d13dea4371f (github-personal/merge, merge)
|\ \  Merge: e5847718e3 d359e99c38
| |/  Author: Matthew Ahrens <mahrens@delphix.com>
| |   Date:   Mon Sep 30 09:10:05 2019 -0700
| |
| |       Merge commit 'd359e99c38f66732d42278c32d52cfcf1839aa4f' into merge
```

IMO, this is overkill for a 2-line commit, especially considering that the original commit doesn't apply cleanly after the refactor, so I had to copy/paste it back in anyway (I couldn't just `git cherry-pick` it, since it tries to apply it to zvol.c, and I couldn't `patch` it to zvol_os.c since it doesn't apply cleanly).  But it is kind of neat that this double-merge works.

As for adding a comment about `zvol_fake_phys_block_size`, I can do that easily enough but I don't think that's really in the spirit of a merge commit.  Maybe I'll throw it in as another commit in the branch that I'm merging.

ahrens commented:
@pzakha OK I updated this to have the double-merge structure mentioned above, and added the comment you requested.  Here are all the commits from master:

```
*   commit e1b64c6206b1fd33d6aa0bba5f064604dea6eead
|\  Merge: b0230ebc35 36583fdf0d
| | Author: Matthew Ahrens <mahrens@delphix.com>
| | Date:   Wed Oct 2 12:39:25 2019 -0700
| | 
| |     Merge branch 'merge-test' into merge2
| | 
| * commit 36583fdf0db839d0c2b050a6f8983c0ec1b9dfdd
| | Author: Matthew Ahrens <mahrens@delphix.com>
| | Date:   Wed Oct 2 12:37:20 2019 -0700
| | 
| |     add comment about zvol_fake_phys_block_size
| | 
| * commit 4c3b17f36f32ab08f7dea092bbd3134a8713294c
| | Author: Matthew Ahrens <mahrens@delphix.com>
| | Date:   Tue Oct 1 14:25:19 2019 -0700
| | 
| |     dx4linux: override phys block size
| |   
| *   commit ef62b60df15e6093309113457e2c76f5fb68d6c9
| |\  Merge: 173c8e3bf8 5df7e9d85c
| | | Author: Matthew Ahrens <mahrens@delphix.com>
| | | Date:   Tue Oct 1 14:16:39 2019 -0700
| | | 
| | |     Merge commit '5df7e9d85ced424292e8312a7848b6092a29c930' into merge-test
| | | 
| | * commit 5df7e9d85ced424292e8312a7848b6092a29c930
| | | Author: Matthew Macy <mmacy@mattmacy.io>
| | | Date:   Wed Sep 25 09:20:30 2019 -0700
| | | 
| | |     OpenZFS restructuring - zvol
| | |     
| | |     Refactor the zvol in to platform dependent and independent bits.
| | |     
| | |     Reviewed-by: Allan Jude <allanjude@freebsd.org>
| | |     Reviewed-by: Jorgen Lundman <lundman@lundman.net>
| | |     Reviewed-by: Igor Kozhukhov <igor@dilos.org>
| | |     Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
| | |     Signed-off-by: Matt Macy <mmacy@FreeBSD.org>
| | |     Closes #9295
| | | 
| * | commit 173c8e3bf83712a46f5978d7defe1dce150cb90a
|/ /  Author: Matthew Ahrens <mahrens@delphix.com>
| |   Date:   Tue Oct 1 14:16:18 2019 -0700
| |   
| |       Revert "dx4linux: override phys block size"
| |       
| |       This reverts commit 883275865df4495e31585e4d7465a8eac8142857.
| |   
* |   commit b0230ebc3580ff62827b7a7fbfbb6d13dea4371f
|\ \  Merge: e5847718e3 d359e99c38
| |/  Author: Matthew Ahrens <mahrens@delphix.com>
| |   Date:   Mon Sep 30 09:10:05 2019 -0700
| |   
| |       Merge commit 'd359e99c38f66732d42278c32d52cfcf1839aa4f' into merge
| | 
| * commit d359e99c38f66732d42278c32d52cfcf1839aa4f
| | Author: loli10K <loli10K@users.noreply.github.com>
| | Date:   Tue Sep 24 21:01:37 2019 +0200
| | 
| |     diff_cb() does not handle large dnodes
| |     
| |     Trying to 'zfs diff' a snapshot with large dnodes will incorrectly try
| |     to access its interior slots when dnodesize > sizeof(dnode_phys_t).
| |     This is normally not an issue because the interior slots are
| |     zero-filled, which report_dnode() handles calling
| |     report_free_dnode_range(). However this is not the case for encrypted
| |     large dnodes or filesystem using many SA based xattrs where the extra
| |     data past the legacy dnode size boundary is interpreted as a
| |     dnode_phys_t.
| |     
| |     Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
| |     Reviewed-by: Tom Caputi <tcaputi@datto.com>
| |     Reviewed-by: Ryan Moeller <ryan@ixsystems.com>
| |     Signed-off-by: loli10K <ezomori.nozomu@gmail.com>
| |     Closes #7678
| |     Closes #8931
| |     Closes #9343
| | 
| * commit 73d7820bba26dd61085cb6dd8348ebf0bb41b951
| | Author: Ryan Moeller <ryan@freqlabs.com>
| | Date:   Sun Sep 22 18:27:53 2019 -0400
| | 
| |     Use signed types to prevent subtraction overflow
| |     
| |     The difference between the sizes could be positive or negative. Leaving
| |     the types as unsigned means the result overflows when the difference is
| |     negative and removing the labs() means we'll have introduced a bug. The
| |     subtraction results in the correct value when the unsigned integer is
| |     interpreted as a signed integer by labs().
| |     
| |     Clang doesn't see that we're doing a subtraction and abusing the types.
| |     It sees the result of the subtraction, an unsigned value, being passed
| |     to an absolute value function and emits a warning which we treat as an
| |     error.
| |     
| |     Reviewed by: Youzhong Yang <youzhong@gmail.com>
| |     Reviewed-by: Igor Kozhukhov <igor@dilos.org>
| |     Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
| |     Signed-off-by: Ryan Moeller <ryan@ixsystems.com>
| |     Closes #9355
| | 
| * commit d49d7336dd42d19de00ee24d1112d496829ff027
| | Author: Kody A Kantor <kody.kantor@gmail.com>
| | Date:   Sun Sep 22 17:25:39 2019 -0500
| | 
| |     Disabled resilver_defer feature leads to looping resilvers
| |     
| |     When a disk is replaced with another on a pool with the resilver_defer
| |     feature present, but not enabled the resilver activity restarts during
| |     each spa_sync. This patch checks to make sure that the resilver_defer
| |     feature is first enabled before requesting a deferred resilver.
| |     
| |     This was originally fixed in illumos-joyent as OS-7982.
| |     
| |     Reviewed-by: Chris Dunlop <chris@onthe.net.au>
| |     Reviewed-by: George Melikov <mail@gmelikov.ru>
| |     Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
| |     Reviewed-by: Tom Caputi <tcaputi@datto.com>
| |     Reviewed by: Jerry Jelinek <jerry.jelinek@joyent.com>
| |     Signed-off-by: Kody A Kantor <kody@kkantor.com>
| |     External-issue: illumos-joyent OS-7982
| |     Closes #9299
| |     Closes #9338
| | 
| * commit afc8f0a6ffb4dd2dd5e17abc39e035eb7c7bcdc8
| | Author: Ryan Moeller <ryan@freqlabs.com>
| | Date:   Wed Sep 18 12:05:57 2019 -0400
| | 
| |     Refactor libzfs_error_init newlines
| |     
| |     Move the trailing newlines from the error message strings to the format
| |     strings to more closely match the other error messages.
| |     
| |     Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
| |     Reviewed-by: George Melikov <mail@gmelikov.ru>
| |     Signed-off-by: Ryan Moeller <ryan@ixsystems.com>
| |     Closes #9330
| | 
| * commit dd262c9681793e2bae6528297803e627622801f0
| | Author: Andriy Gapon <avg@FreeBSD.org>
| | Date:   Wed Sep 18 19:04:45 2019 +0300
| | 
| |     Fix dsl_scan_ds_clone_swapped logic
| |     
| |     The was incorrect with respect to swapping dataset IDs both in the
| |     on-disk ZAP object and the in-memory queue.
| |     
| |     In both cases, if ds1 was already present, then it would be first
| |     replaced with ds2 and then ds would be replaced back with ds1.
| |     Also, both cases did not properly handle a situation where both ds1 and
| |     ds2 are already queued.  A duplicate insertion would be attempted and
| |     its failure would result in a panic.
| |     
| |     Reviewed-by: Matt Ahrens <matt@delphix.com>
| |     Reviewed-by: Tom Caputi <tcaputi@datto.com>
| |     Signed-off-by: Andriy Gapon <avg@FreeBSD.org>
| |     Closes #9140
| |     Closes #9163
| | 
* | commit e5847718e356846129c7ce2e9327c60bfd9cbb1f (origin/master)
```
